src/pubsub/ua_pubsub_reader.c

```diff
@@ -929,7 +933,7 @@ UA_DataSetReader_process(UA_Server *server, UA_ReaderGroup *rg,
             if(tv->beforeWrite)
                 tv->beforeWrite(server, &dsr->identifier, &dsr->linkedReaderGroup,
                                 &tv->targetVariable.targetNodeId,
-                                tv->targetVariableContext, tv->externalDataValue);
+                                tv->targetVariableContext, &field);
```

Comment:
I think, there is a misunderstanding.
We want to give `tv->externalDataValue` into beforeWrite, so that the target memory location can be switched out before writing into it.
So that we can do "double buffering" where we don't want to have second thread work on a half-updated value.

Reply:
I tried to fix this with an updated commit similar to the first call of beforeWrite. Is this ok this way?

Reply:
@jpfr: If the solution in this PR is not suitable for the double buffering, please provide a solution where tv->externalDataValue is provided with the new value to beforeWrite according to the documentation. The user callbacks rely on the documentation.

Reply:
Set `*tv->externalDataValue` to the new location so that `**tv->externalDataValue` gets written to it.

Reply:
@jpfr: Set *tv->externalDataValue now.
Is the first change from line 815 on fine?

Reply:
I don't understand the preview thing.
If you are unsure if the data is acceptable, you can use `beforeWrite` to set up a temporary memory (and have the double-pointer point to it) and in `afterWrite` decide whether you want to to keep the value...

Reply:
I don't understand it either. I just noticed the incompatible change in the callback usage.